sklearn/_fast_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,754 @@
+                for estimator in predictors_at_ith_iteration)
+            n_predictors = sum(
+                len(predictors_at_ith_iteration)
+                for predictors_at_ith_iteration in self.estimators_)
+            print("Fit {} trees in {:.3f} s, ({} total leaves)".format(
+                n_predictors, duration, n_total_leaves))
+            print("{:<32} {:.3f}s".format('Time spent finding best splits:',
+                                          acc_find_split_time))
+            print("{:<32} {:.3f}s".format('Time spent applying splits:',
+                                          acc_apply_split_time))
+            print("{:<32} {:.3f}s".format('Time spent predicting:',
+                                          acc_prediction_time))
+
+        self.train_score_ = np.asarray(self.train_score_)
+        self.validation_score_ = np.asarray(self.validation_score_)
+        self._in_fit = False
+        return self
+
+    def _check_early_stopping(self, X_binned_train, y_train,
+                              X_binned_val, y_val):
+        """Check if fitting should be early-stopped.
+
+        Scores are computed on validation data or on training data.
+        """
+
+        self.train_score_.append(
+            self._get_scores(X_binned_train, y_train))
+
+        if self.validation_fraction is not None:
+            self.validation_score_.append(
+                self._get_scores(X_binned_val, y_val))
+            return self._should_stop(self.validation_score_)
+
+        return self._should_stop(self.train_score_)
+
+    def _should_stop(self, scores):
+        """
+        Return True (do early stopping) if the last n scores aren't better
+        than the (n-1)th-to-last score, up to some tolerance.
+        """
+        reference_position = self.n_iter_no_change + 1
+        if len(scores) < reference_position:
+            return False
+
+        # A higher score is always better. Higher tol means that it will be
+        # harder for subsequent iteration to be considered an improvement upon
+        # the reference score, and therefore it is more likely to early stop
+        # because of the lack of significant improvement.
+        tol = 0 if self.tol is None else self.tol
+        reference_score = scores[-reference_position] + tol
+        recent_scores = scores[-reference_position + 1:]
+        recent_improvements = [score > reference_score
+                               for score in recent_scores]
+        return not any(recent_improvements)
+
+    def _get_scores(self, X, y):
+        """Compute scores on data X with target y.
+
+        Scores are computed with a scorer if scoring parameter is not
+        'loss', else with the loss. As higher is always better, we return
+        -loss_value.
+        """
+
+        if self.scoring != 'loss':
+            return self.scorer_(self, X, y)
+
+        # Else, use loss
+        raw_predictions = self._raw_predict(X)
```

Comment:
We could pass an `self._raw_predict(X, assume_binned=True)` here to get rid of the `self._in_fit = True` stateful hack.

Reply:
Alternatively we could just always pass the unbinned data to `_raw_predict`, both during `fit` and `predict` as there is no real computation advantage to predicting on binned feature.

The only problem is the  extra `train_test_split` masking on the unbinned data that will cause a dataset copy for nothing (for the training part but it can be resolved such as:

```python
            X_binned_train, X_binned_val, y_train, y_val, _, idx_val = train_test_split(
                X_binned, y, np.arange(y.shape[0]), test_size=self.validation_fraction,
                stratify=stratify, random_state=rng)
            X_val = sage_indexing(X, idx_val)
```

Reply:
>there is no real computational benefit to predicting on binned feature

Wait, the reason we need all this is precisely because predicting on binned data is faster then predicting on numeric data, right?

Reply:
We would have to check but I thought it wasn't necessarily the case.

Reply:
https://github.com/ogrisel/pygbm/issues/63#issue-389478701

It was faster on pygbm. I'll run the same benchmark here to be sure